pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/Node.java

```diff
@@ -315,10 +399,48 @@
      *
      * @param childIndex
      *          The index of the child to be removed
+     *
+     * @deprecated This is internal API and will be removed from this interface with 7.0.0
      */
+    @Deprecated
+    @InternalApi
     void removeChildAtIndex(int childIndex);
 
 
+    /**
+     * Returns the parent of this node, or null if this is the {@linkplain RootNode root}
+     * of the tree.
+     *
+     * <p>This method should be preferred to {@link #jjtGetParent()}.
+     *
+     * @return The parent of this node
+     */
+    Node getParent();
+
+    /**
+     * Returns the child of this node at the given index.
+     *
+     * @throws IndexOutOfBoundsException if the index is negative or greater than {@link #getNumChildren()}.
+     */
+    Node getChild(int index);
```

Comment:
Should we add a method `List<Node> getChildren()` instead of `getChild(int)` and `getNumChildren()`?
It would make the API a bit more secure (e.g. you can't ask for a out of bounds child) and on this level there would be only one method instead of two.
However, if one wants to know the number of children, then it would be always a chained call like `getChildren().size()`...
What do you think?

Reply:
With the node stream API I'm not sure this matters much. There's a method `NodeStream<Node> children()` on the node interface, and to get a list you can do `children().toList()` (which IIRC might predimension the list too). If we have to choose I think I prefer the NodeStream there.

Reply:
ok, let's add the getParent() and we'll have both ways. As you mentioned, the NodeStream API is not always the best choice, if you only want e.g. the first parent.